notification/src/api/notification/notification.controller.js

```diff
@@ -0,0 +1,133 @@
+async function handleNotification (request, response, logger) {
+  const body = await httpUtils.collectRequestData(request, response)
+  try {
+    await processNotifications(JSON.parse(body), logger)
+    response.statusCode = 200
+    return response.end("{ notificationResponse : '[accepted]' }")
+  } catch(e) {
+    logger.error(e, 'Ooops')
+    response.statusCode = 500
+    return response.end()
+  }
+}
+
+// TODO: proper naming
+async function processNotifications (request, logger) {
+  const notifications = request.notificationItems
+  await Promise.map(notifications, notification => processNotification(notification, logger))
+}
+
+// TODO: proper naming
+async function processNotification (notification, logger) {
+  const merchantReference = _.get(notification, 'NotificationRequestItem.merchantReference', null)
+  if (merchantReference === null) {
+    logger.error(`Can't extract merchantReference from the notification: ${JSON.stringify(notification)}`)
+    return null
+  }
+
+  let updateActions
+  let payment
+  try {
+    payment = await getPaymentByMerchantReference(merchantReference)
+    updateActions = calculateUpdateActionsForPayment(payment, notification)
+  } catch (e) {
+    logger.warn(e, `Error while trying to fetch the payment for merchantReference: ${merchantReference}`)
+    return null
+  }
+
+  try {
+    await ctpClient.update(ctpClient.builder.payments, payment.id, payment.version, updateActions)
+  } catch (e) {
+    logger.error(e, `Error while updating the payment with id: ${payment.id}`)
+  }
+}
+
+function calculateUpdateActionsForPayment (payment, notification) {
+  const notificationEventCode = notification.eventCode
+  const notificationSuccess = notification.success
+  const updateActions = []
+  const stringifiedNotification = JSON.stringify(notification)
+  // check if interfaceInteraction already on payment or not
+  const isInterfaceInteractionOnPayment =
+    payment.interfaceInteractions.some(interaction => interaction.fields.response === stringifiedNotification)
+  if (isInterfaceInteractionOnPayment === false)
+    updateActions.push(getAddInterfaceInteractionUpdateAction(notification))
+
+  const { transactionType, transactionState } = getTransactionTypeAndStateUpdateAction(notificationEventCode, notificationSuccess)
+  if (transactionType !== null) {
+    // if there is already a transaction with type `transactionType` then update its `transactionState` if necessary,
+    // otherwise create a transaction with type `transactionType` and state `transactionState`
+    const oldTransaction = _.find(payment.transactions, (transaction) => transaction.type === transactionType)
+    if (_.isEmpty(oldTransaction)) {
+      updateActions.push(getAddTransactionUpdateAction(
+        transactionType,
+        transactionState,
+        notification.amount.value,
+        notification.amount.currency)
+      )
+    } else if (ctp.compareTransactionStates(oldTransaction.state, transactionState)) {
```

Comment:
Don't you want to be more explicit here? `compareTransactionStates` doesn't return true/false, so maybe compare it to number, e.g. `if (ctp.compareTransactionStates() !== 0)`

Reply:
https://github.com/commercetools/commercetools-adyen-integration/commit/2d5c30a4cac5659c582128307462a802340e5c11